File: tests/unit/test/annotation.cpp
#include "test/annotation.h"

#include <cctype>

#include "llvm/ADT/STLExtras.h"
#include "llvm/ADT/SmallVector.h"

namespace clice::testing {

/// The annotation grammar reserves no ASCII character: the sigil is `§`
/// (U+00A7) and range bodies are delimited by `⟦` / `⟧` (U+27E6 / U+27E7),
/// none of which occur in real C++, Doxygen, Markdown, CMake or LSP
/// snippet content. ASCII parentheses are safe for names because they only
/// have meaning directly after the sigil:
///
///   §              nameless point       §(name)          named point
///   §⟦...⟧         nameless range       §(name)⟦...⟧     named range
///
/// Ranges may nest. Everything else — Doxygen tags, array subscripts,
/// `${1:...}` snippet placeholders — is ordinary source text and passes
/// through untouched. A stray `⟦` or `⟧` is a hard error so that typos
/// fail loudly instead of silently shifting offsets.
AnnotatedSource AnnotatedSource::from(llvm::StringRef content) {
    std::string source;
    source.reserve(content.size());

    llvm::StringMap<std::uint32_t> offsets;
    llvm::StringMap<LocalSourceRange> ranges;
    std::vector<std::uint32_t> nameless_offsets;

    constexpr llvm::StringRef sigil = "§";
    constexpr llvm::StringRef open_mark = "⟦";
    constexpr llvm::StringRef close_mark = "⟧";

    struct OpenSpan {
        llvm::StringRef key;
        std::uint32_t begin;
    };

    llvm::SmallVector<OpenSpan, 4> stack;

    std::uint32_t offset = 0;
    std::uint32_t i = 0;

    while(i < content.size()) {
        llvm::StringRef rest = content.substr(i);

        if(rest.starts_with(sigil)) {
            i += sigil.size();

            llvm::StringRef key;
            if(i < content.size() && content[i] == '(') {
                std::size_t key_end = content.find(')', i + 1);
                assert(key_end != llvm::StringRef::npos && "Unterminated §(name).");
                key = content.slice(i + 1, key_end);
                // A name must be a plain identifier (or empty): `§` directly
                // before a real parenthesized expression is ambiguous — write
                // the explicit nameless form `§()` in front of it instead.
                assert(llvm::all_of(key,
                                    [](char c) {
                                        return std::isalnum(static_cast<unsigned char>(c)) ||
                                               c == '_';
                                    }) &&
                       "§(name) requires an identifier name; use §() for a nameless point.");
                i = static_cast<std::uint32_t>(key_end) + 1;
            }

            if(content.substr(i).starts_with(open_mark)) {
                stack.push_back({key, offset});
                i += open_mark.size();
            } else if(key.empty()) {
                nameless_offsets.emplace_back(offset);
            } else {
                offsets.try_emplace(key, offset);
            }
            continue;
        }

        if(rest.starts_with(close_mark)) {
            assert(!stack.empty() && "`⟧` without a matching `§⟦`.");
            OpenSpan span = stack.pop_back_val();
            ranges.try_emplace(span.key, LocalSourceRange{span.begin, offset});
            i += close_mark.size();
            continue;
        }

        assert(!rest.starts_with(open_mark) && "`⟦` must follow `§` or `§(name)`.");

        source += content[i];
        offset += 1;
        i += 1;
    }

    assert(stack.empty() && "Unclosed `§⟦` annotation at end of input.");

    return AnnotatedSource{
        std::move(source),
        std::move(offsets),
        std::move(ranges),
        std::move(nameless_offsets),
    };
}

void AnnotatedSources::add_sources(llvm::StringRef content) {
    std::string curr_file;
    std::string curr_content;

    auto save_previous_file = [&]() {
        if(curr_file.empty()) {
            return;
        }

        add_source(curr_file, curr_content);
        curr_file.clear();
        curr_content.clear();
    };

    while(!content.empty()) {
        llvm::StringRef line = content.take_front(content.find_first_of("\r\n"));
        content = content.drop_front(line.size());
        if(content.starts_with("\r\n")) {
            content = content.drop_front(2);
        } else if(content.starts_with("\n")) {
            content = content.drop_front(1);
        }

        if(line.starts_with("#[") && line.ends_with("]")) {
            save_previous_file();
            curr_file = line.slice(2, line.size() - 1).str();
        } else if(!curr_file.empty()) {
            curr_content += line;
            curr_content += '\n';
        }
    }

    save_previous_file();
}

}  // namespace clice::testing

File: tests/tools/annotation.py
"""Annotated-source parser mirroring tests/unit/test/annotation.cpp.

Fixture files may carry inline annotations: `§` / `§(name)` points and
`§⟦...⟧` / `§(name)⟦...⟧` ranges (nesting allowed). The unit Tester strips
them before compiling; the integration harness strips them before didOpen,
so the server compiles the same clean text while the test keeps the marked
offsets. Files without annotations pass through unchanged.

Offsets are byte offsets in the stripped UTF-8 text, matching the C++ side.
"""

from dataclasses import dataclass, field

SIGIL = "§".encode()
OPEN_MARK = "⟦".encode()
CLOSE_MARK = "⟧".encode()


@dataclass
class AnnotatedSource:
    content: str
    offsets: dict[str, int] = field(default_factory=dict)
    ranges: dict[str, tuple[int, int]] = field(default_factory=dict)
    nameless_offsets: list[int] = field(default_factory=list)


def parse_annotations(text: str) -> AnnotatedSource:
    data = text.encode()
    source = bytearray()
    offsets: dict[str, int] = {}
    ranges: dict[str, tuple[int, int]] = {}
    nameless: list[int] = []
    stack: list[tuple[str, int]] = []

    i = 0
    while i < len(data):
        if data.startswith(SIGIL, i):
            i += len(SIGIL)

            key = ""
            if i < len(data) and data[i : i + 1] == b"(":
                key_end = data.find(b")", i + 1)
                if key_end < 0:
                    raise ValueError("Unterminated §(name) in annotated source.")
                key = data[i + 1 : key_end].decode()
                # ASCII-only like the C++ twin's std::isalnum; str.isalnum
                # would accept non-ASCII letters the unit Tester rejects.
                if not all(c.isascii() and (c.isalnum() or c == "_") for c in key):
                    raise ValueError(
                        f"§({key}) is not an identifier name; use §() for a "
                        "nameless point before real parentheses."
                    )
                i = key_end + 1

            if data.startswith(OPEN_MARK, i):
                stack.append((key, len(source)))
                i += len(OPEN_MARK)
            elif not key:
                nameless.append(len(source))
            elif key in offsets:
                raise ValueError(f"Duplicate point annotation §({key}).")
            else:
                offsets[key] = len(source)
            continue

        if data.startswith(CLOSE_MARK, i):
            if not stack:
                raise ValueError("`⟧` without a matching `§⟦` in annotated source.")
            key, begin = stack.pop()
            if key in ranges:
                raise ValueError(
                    f"Duplicate range annotation §({key})⟦...⟧."
                    if key
                    else "Multiple nameless ranges in one source; name them instead."
                )
            ranges[key] = (begin, len(source))
            i += len(CLOSE_MARK)
            continue

        if data.startswith(OPEN_MARK, i):
            raise ValueError("`⟦` must follow `§` or `§(name)`.")

        source.append(data[i])
        i += 1

    if stack:
        raise ValueError("Unclosed `§⟦` annotation at end of input.")

    return AnnotatedSource(source.decode(), offsets, ranges, nameless)
